Wasm Configuration for Ambient Mode (#15041)
istio/istio.io: content/en/docs/ambient/usage/extend-waypoint-wasm/index.md

---

> title: Distributing WebAssembly Modules

Reviewer: The title was supposed to be "Extend waypoints with WebAssembly plugins" ?

Author: corrected

description: Describes how to make remote WebAssembly modules available for ambient mode.
weight: 45
keywords: [extensibility,Wasm,WebAssembly,Ambient]

> owner: istio/wg-policies-and-telemetry-maintainers

Reviewer 2: an odd owner

Reviewer: All wasm documents are currently under this ownership, not sure why.. so did not change.

test: no
status: Alpha
---

{{< boilerplate alpha >}}

> Istio provides the ability to [extend its functionality using WebAssembly (Wasm)](/docs/reference/config/proxy_extensions/wasm-plugin/).

Reviewer 2: `/docs/concepts/wasm`

Author: corrected

> One of the key advantages of Wasm extensibility is that extensions can be loaded dynamically at runtime. This document outlines the testing process for the implementation of Wasm features in ambient mode within Istio. In ambient mode, Wasm configuration must be applied to the waypoint proxy deployed in each namespace, instead of to individual sidecars.

Reviewer 2: Why is this about testing, rather than using?

## Install Ambient Mode and deploy test applications

> Follow the [Ambient Getting Started Guide](docs/ambient/getting-started/#download) to install Istio in ambient mode. Deploy the [sample applications](docs/ambient/getting-started/#bookinfo) required for exploring waypoint proxy extensibility via Wasm. Make sure to add the sample [applications to ambient mode](docs/ambient/getting-started/#addtoambient) before proceeding further.

Reviewer 2: ```suggestion
Follow the [Ambient Getting Started Guide](docs/ambient/getting-started/#download) to install Istio in ambient mode. Deploy the [sample applications](docs/ambient/getting-started/#bookinfo) required for exploring waypoint proxy extensibility via Wasm. Make sure to [add the sample applications](docs/ambient/getting-started/#addtoambient) to the mesh before proceeding further.
```

Author: changed

## Apply Wasm configuration at the Gateway

> 

Reviewer 2: some text between headings please

Author: added

  url: oci://ghcr.io/istio-ecosystem/wasm-extensions/basic_auth:1.12.0
  phase: AUTHN
  pluginConfig:
    basic_auth_rules:
      - prefix: "/productpage"
        request_methods:
          - "GET"
          - "POST"
        credentials:
          - "ok:test"
          - "YWRtaW4zOmFkbWluMw=="
EOF
{{< /text >}}

An HTTP filter will be injected at the gateway as an authentication filter.
The Istio agent will interpret the WasmPlugin configuration, download remote Wasm modules from the OCI image registry to a local file, and inject the HTTP filter at the gateway by referencing that file.

### Verify the traffic via the Gateway

1. Test `/productpage` without credentials

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null "http://$GATEWAY_HOST/productpage"
    401
    {{< /text >}}

1. Test `/productpage` with credentials configured in the WasmPlugin resource

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -o /dev/null -H "Authorization: Basic YWRtaW4zOmFkbWluMw==" -w "%{http_code}" "http://$GATEWAY_HOST/productpage"
    200
    {{< /text >}}

## Apply Wasm Configuration at Waypoint Proxy

> 

Reviewer 2: and here

Author: added

EOF
{{< /text >}}

### Verify the traffic targeting the Service

1. Test internal `/productpage` with credentials configured at the generic `waypoint` proxy

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null -H "Authorization: Basic YWRtaW4zOmFkbWluMw==" http://productpage:9080/productpage
    200
    {{< /text >}}

1. Test internal `/reviews` with credentials configured at the specific `reviews-svc-waypoint` proxy

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null -H "Authorization: Basic MXQtaW4zOmFkbWluMw==" http://reviews:9080/reviews/1
    200
    {{< /text >}}

1. Test internal `/productpage` without credentials

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null http://reviews:9080/reviews/1
    401
    {{< /text >}}

When executing the provided command without credentials, it verifies that accessing the internal `/productpage` results in a 401 unauthorized response, demonstrating the expected behavior of failing to access the resource without proper authentication credentials.

### Cleanup

{{< text bash >}}

> $ kubectl delete wasmplugin basic-auth-at-gateway basic-auth-at-waypoint basic-auth-for-service

Reviewer 3: Add a clean up for waypoint? and remove the use-waypoint label?

Reviewer: @linsun : i think it is not needed, as at line 240 we refer to ambient cleanup document which includes steps to clean up which already includes these steps? https://preliminary.istio.io/latest/docs/ambient/getting-started/#uninstall

Reviewer 3: Oh sounds good, thx for the clarification!

Reviewer 2: please add some descriptive text to this section above the command (look at how it's done elsewhere if you need)

{{< /text >}}

> Follow [ambient uninstall guide](docs/ambient/getting-started/#uninstall) for cleanup of ambient mode and sample test applications.

Reviewer 2: ```suggestion
Follow [the ambient mode uninstall guide](docs/ambient/getting-started/#uninstall) to remove Istio and sample test applications.
```

Nit: I don't like our docs telling people to uninstall Istio!

Author: changed

description: Describes how to make remote WebAssembly modules available for ambient mode.
weight: 55
keywords: [extensibility,Wasm,WebAssembly,Ambient]
owner: istio/wg-policies-and-telemetry-maintainers
test: no
status: Alpha
---

{{< boilerplate alpha >}}

Istio provides the ability to [extend its functionality using WebAssembly (Wasm)](/docs/concepts/wasm/).
One of the key advantages of Wasm extensibility is that extensions can be loaded dynamically at runtime. This document outlines how to extend ambient mode within Istio with Wasm features. In ambient mode, Wasm configuration must be applied to the waypoint proxy deployed in each namespace, instead of to individual sidecars.

## Install Ambient Mode and deploy test applications

Follow the [Ambient Getting Started Guide](docs/ambient/getting-started/#download) to install Istio in ambient mode. Deploy the [sample applications](docs/ambient/getting-started/#bookinfo) required for exploring waypoint proxy extensibility via Wasm. Make sure to [add the sample applications](docs/ambient/getting-started/#addtoambient) to the mesh before proceeding further.

## Apply Wasm configuration at the Gateway

> Istio's Gateway provides a centralized entry point for managing traffic into the service mesh. Here, we'll configure a WasmPlugin to enhance security by adding HTTP Basic authentication. This plugin will be applied at the gateway level, ensuring that all traffic passing through the gateway is subject to the configured authentication rules.

Reviewer: ```suggestion
With Kubernetes Gateway API, Istio provides a centralized entry point for managing traffic into the service mesh. We will configure a WasmPlugin at the gateway level, ensuring that all traffic passing through the gateway is subject to the extended authentication rules.
```

Author: changed

  url: oci://ghcr.io/istio-ecosystem/wasm-extensions/basic_auth:1.12.0
  phase: AUTHN
  pluginConfig:
    basic_auth_rules:
      - prefix: "/productpage"
        request_methods:
          - "GET"
          - "POST"
        credentials:
          - "ok:test"
          - "YWRtaW4zOmFkbWluMw=="
EOF
{{< /text >}}

An HTTP filter will be injected at the gateway as an authentication filter.
The Istio agent will interpret the WasmPlugin configuration, download remote Wasm modules from the OCI image registry to a local file, and inject the HTTP filter at the gateway by referencing that file.

### Verify the traffic via the Gateway

1. Test `/productpage` without credentials

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null "http://$GATEWAY_HOST/productpage"
    401
    {{< /text >}}

1. Test `/productpage` with credentials configured in the WasmPlugin resource

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -o /dev/null -H "Authorization: Basic YWRtaW4zOmFkbWluMw==" -w "%{http_code}" "http://$GATEWAY_HOST/productpage"
    200
    {{< /text >}}

## Apply Wasm Configuration at Waypoint Proxy

> Waypoint proxies play a crucial role in Istio's ambient mode, facilitating secure and efficient communication within the service mesh. Below, we'll explore how to apply Wasm configuration to these proxies, enhancing their functionality dynamically.

Reviewer: ```suggestion
Waypoint proxies play a crucial role in Istio's ambient mode, facilitating secure and efficient communication within the service mesh. Below, we will explore how to apply Wasm configuration to the waypoint, enhancing the proxy functionality dynamically.
```

Author: changed

          - "MXQtaW4zOmFkbWluMw=="
EOF
{{< /text >}}

### Verify the traffic targeting the Service

1. Test internal `/productpage` with credentials configured at the generic `waypoint` proxy

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null -H "Authorization: Basic YWRtaW4zOmFkbWluMw==" http://productpage:9080/productpage
    200
    {{< /text >}}

1. Test internal `/reviews` with credentials configured at the specific `reviews-svc-waypoint` proxy

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null -H "Authorization: Basic MXQtaW4zOmFkbWluMw==" http://reviews:9080/reviews/1
    200
    {{< /text >}}

1. Test internal `/productpage` without credentials

    {{< text bash >}}
    $ kubectl exec deploy/sleep -- curl -s -w "%{http_code}" -o /dev/null http://reviews:9080/reviews/1
    401
    {{< /text >}}

When executing the provided command without credentials, it verifies that accessing the internal `/productpage` results in a 401 unauthorized response, demonstrating the expected behavior of failing to access the resource without proper authentication credentials.

### Cleanup

> When you're finished experimenting with the Wasm plugins, uninstall and clean them using the following command.

Reviewer: 1. Remove WasmPlugin configuration:

> When you're finished experimenting with the Wasm plugins, uninstall and clean them using the following command.

Reviewer: ```suggestion
1. Remove WasmPlugin configuration:
```

Author: corrected

{{< text bash >}}
$ kubectl delete wasmplugin basic-auth-at-gateway basic-auth-at-waypoint basic-auth-for-service
{{< /text >}}

> Follow [the ambient mode uninstall guide](docs/ambient/getting-started/#uninstall) to remove Istio and sample test applications.

Reviewer: ```suggestion
1. Follow [the ambient mode uninstall guide](docs/ambient/getting-started/#uninstall) to remove Istio and sample test applications.
```

Author: corrected
